src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php

```diff
@@ -3279,6 +3291,14 @@ private static function getTypedValueEqualityAssertions(
             } else {
                 $if_types[$var_name] = [['~' . $var_type->getAssertionString()]];
             }
+
+            if ($other_var_name && $other_type && count($other_type->getAtomicTypes()) === 1) {
```

Comment:
A comment stating why `Union::isSingle()` is not used here could be useful.

Reply:
Note: the following does not work even when `isSingle()` is used.
```php
/** @var 1|2 $a */
$a = 2;

/** @var 2|3 $b */
$b = 2;

if ($a === $b) {
    /** @psalm-trace $a */; // expected: 2, actual: 2
    /** @psalm-trace $b */; // expected: 2, actual: 2|3
}
```

Reply:
Yeah, that's exactly why I count atomics instead. When the Union contains multiple literal of the same type, isSingle still returns true but getAssertionString will return the assertion only for the first Atomic. I'm adding the comment